### PR TITLE
Device Manager

### PR DESCRIPTION
**Pull request by dadgar**
Introduce a device manager that manages the lifecycle of device plugins
on the client. It fingerprints, collects stats, and forwards Reserve
requests to the correct plugin. The manager, also handles device plugins
failing and validates their output.

### CODE REVIEW

File: client/devicemanager/instance.go [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+
+	return
+}
+
+// handleFingerprint stores the new devices and triggers the fingerprint output
+// channel. An error is returned if the passed devices don't pass validation.
+func (i *instanceManager) handleFingerprint(f *device.FingerprintResponse) error {
+	// Safety check
+	if f.Devices == nil {
+		return nil
```
Comment (schmichael) [inline client/devicemanager/instance.go]: Should this ever happen? If it's a programming error we should at least log so developers aren't wondering why nothing is happening if they forget to set that.

File: client/devicemanager/instance.go [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+	var ok bool
+	for {
+		select {
+		case <-i.ctx.Done():
+			return
+		case sresp, ok = <-statsCh:
+		}
+
+		if !ok {
+			i.logger.Debug("exiting since stats gracefully shutdown")
```
Comment (schmichael) [inline client/devicemanager/instance.go]: Same comment as above: lets Trace log at most on happy shutdown paths. Our logs are getting really noisy.

File: client/devicemanager/manager.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+	// state.
+	GetDevicePluginState() (*state.PluginState, error)
+
+	// PutDevicePluginState is used to store the device manager's plugin
+	// state.
+	PutDevicePluginState(state *state.PluginState) error
+}
+
+// UpdateNodeDevices is a callback for updating the set of devices on a node.
+type UpdateNodeDevices func(devices []*structs.NodeDeviceResource)
```
Comment (schmichael) [inline client/devicemanager/manager.go]: Lets either use the Fn suffix with all func types or none.

File: client/state/state_database.go [hunk 4]
```diff
@@ -36,6 +40,14 @@ var (
...
 	taskLocalStateKey = []byte("local_state")
 	taskStateKey      = []byte("task_state")
+
+	// devManagerBucket is the bucket name containing all device manager related
+	// data
+	devManagerBucket = []byte("devicemanager")
+
+	// devManagerPluginStateKey is the key serialized device manager
+	// plugin state is stored at
+	devManagerPluginStateKey = []byte("plugin-state")
```
Comment (schmichael) [inline client/state/state_database.go]: underscore to match other bucket names
```suggestion
	devManagerPluginStateKey = []byte("plugin_state")
```

File: command/agent/command.go [hunk 5]
```diff
@@ -118,6 +118,7 @@ func (c *Command) readConfig() *Config {
 	}), "consul-auto-advertise", "")
 	flags.StringVar(&cmdConfig.Consul.CAFile, "consul-ca-file", "", "")
 	flags.StringVar(&cmdConfig.Consul.CertFile, "consul-cert-file", "", "")
+	flags.StringVar(&cmdConfig.Consul.KeyFile, "consul-key-file", "", "")
```
Comment (schmichael) [inline command/agent/command.go]: A bit random but good catch :)

File: plugins/device/plugin.go [hunk 6]
```diff
@@ -28,7 +28,8 @@ func (p *PluginDevice) GRPCServer(broker *plugin.GRPCBroker, s *grpc.Server) err
 
 func (p *PluginDevice) GRPCClient(ctx context.Context, broker *plugin.GRPCBroker, c *grpc.ClientConn) (interface{}, error) {
 	return &devicePluginClient{
-		client: proto.NewDevicePluginClient(c),
+		doneCtx: ctx,
```
Comment (nickethier) [inline plugins/device/plugin.go]: Ok so `doneCtx` is closed if/when the client connection closes?

Out of scope for this PR I think, but I think we might need to discuss how (or if we still want) to implement the Shutdown RPC in the base plugin. As it has implications here I think

File: client/client.go [hunk 7]
```diff
@@ -306,6 +320,20 @@ func NewClient(cfg *config.Config, consulCatalog consul.CatalogAPI, consulServic
 		return nil, fmt.Errorf("fingerprinting failed: %v", err)
 	}
 
+	// Setup the device manager
+	devConfig := &devicemanager.Config{
+		Logger:       c.logger,
+		Loader:       c.configCopy.PluginSingletonLoader,
+		PluginConfig: c.configCopy.NomadPluginConfig(),
+		Updater: func(devs []*structs.NodeDeviceResource) {
+			c.logger.Debug("Device Updater called", "count", len(devs), "devices", hclog.Fmt("% #v", pretty.Formatter(devs)))
```
[the remark this answers is not in the document]
Comment (dadgar) [reply client/client.go]: A follow up PR will deal with that

File: client/devicemanager/instance.go (hunk 1 above)
Comment (dadgar) [reply client/devicemanager/instance.go]: Put a better comment. You could imagine a fingerprint returning no devices if there are none on the node.

**Pull request merged**